overlord/storestate/storestate.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/store"
+)
+
+var storeNew = store.New
+
+// StoreState holds the state for the store in the system.
+type StoreState struct {
+	// StoreLocation is the store API's base location.
+	StoreLocation string `json:"store-location"`
+}
+
+// StoreLocation returns the store API's explicit location.
+func StoreLocation(st *state.State) string {
+	var storeState StoreState
+
+	err := st.Get("store", &storeState)
+	if err != nil {
+		return ""
+	}
+
+	return storeState.StoreLocation
+}
+
+// updateStoreLocation updates the store API's location in persistent state.
+func updateStoreLocation(st *state.State, location string) {
+	var storeState StoreState
+	st.Get("store", &storeState)
+	storeState.StoreLocation = location
+	st.Set("store", &storeState)
+}
+
+// A StoreService can find, list available updates and download snaps.
+type StoreService interface {
+	SnapInfo(spec store.SnapSpec, user *auth.UserState) (*snap.Info, error)
+	Find(search *store.Search, user *auth.UserState) ([]*snap.Info, error)
+	LookupRefresh(*store.RefreshCandidate, *auth.UserState) (*snap.Info, error)
+	ListRefresh([]*store.RefreshCandidate, *auth.UserState) ([]*snap.Info, error)
+	Sections(user *auth.UserState) ([]string, error)
+	Download(context.Context, string, string, *snap.DownloadInfo, progress.Meter, *auth.UserState) error
+
+	Assertion(assertType *asserts.AssertionType, primaryKey []string, user *auth.UserState) (asserts.Assertion, error)
+
+	SuggestedCurrency() string
+	Buy(options *store.BuyOptions, user *auth.UserState) (*store.BuyResult, error)
+	ReadyToBuy(*auth.UserState) error
+}
+
+// SetupStore configures the system's initial store.
+func SetupStore(st *state.State, authContext auth.AuthContext) error {
+	storeConfig, err := initialStoreConfig(st)
+	if err != nil {
+		return err
+	}
+	sto := storeNew(storeConfig, authContext)
+	replaceAuthContext(st, authContext)
+	ReplaceStore(st, sto)
+	return nil
+}
+
+// SetStoreLocation replaces the location of the store API used by the system.
+// If the URL is nil the store is reverted to the system's default.
+func SetStoreLocation(state *state.State, u *url.URL) error {
+	location := ""
+	config := store.DefaultConfig()
+	if u != nil {
+		location = u.String()
+		err := config.SetAPI(u)
+		if err != nil {
+			return err
+		}
+	}
+	store := store.New(config, cachedAuthContext(state))
+	ReplaceStore(state, store)
+	updateStoreLocation(state, location)
+	return nil
+}
+
+func initialStoreConfig(st *state.State) (*store.Config, error) {
+	config := store.DefaultConfig()
+	if location := StoreLocation(st); location != "" {
+		u, err := url.Parse(location)
+		if err != nil {
+			return nil, fmt.Errorf("invalid store API location: %s", err)
+		}
+		err = config.SetAPI(u)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return config, nil
+}
+
+type cachedAuthContextKey struct{}
+
+func replaceAuthContext(state *state.State, authContext auth.AuthContext) {
```

Comment:
given that this one is internal  s/replace/cache/ or save|set?

Reply:
I don't mind the current name, nor would I mind save* or set*; I'd object to cache* (because, cachedAuthContextKey vs cacheAuthContext)

Reply:
ReplaceStore is called that way because it's public use is for tests,  stashAuthContext or saveAuthContext would work for me

Reply:
Changed to saveAuthContext.